Коммит доработок по третьему спринту: Роутинг в React приложении. Реализация механизмов авторизации в React приложении.

EldenLordCommander opened the pull request:
Добрый день. Сдаю на проверку мои доработки по третьему спринту. Реализовал весь функционал по ТЗ, протестировал, хотелось бы узнать - правильно ли понял задачи, их реализацию, каким образом должно все работать.

Что реализовано:
- добавил страницы login, register, orders, profile, reseet-password, forgot-password в папку pages в проект и стили к каждой странице
- добавил вызовы каждого апи, описанного в задании, с токенами и без, и с обновлением токена
- добавил экшены, хранилища и редьюсеры для всех вызовов апи используя обычный Redux, с которым решил продолжить работу после прошлого спринта 
- добавил modalswitch для двух режимов работы окна с ингредиентом - как модалка или как отдельная страница, доступная даже неавторизированному юзеру
- добавил роутинг в проект, с редиректами, выделением ссылок, где находится юзер
- добавил ограничение на авторизированных и не авторизированных пользователей на страницы, через компонент и через проверку на наличие авторизированного юзера
- добавил регистрацию, логин, сброс и восстановление пароля для неавторизированного юзера, обновление данных зарегистрированного юзера и логаут

Жду вашей обратной связи по поводу текущей версии моего проекта, спасибо.

EldenLordCommander committed:
Коммит доработок по третьему спринту

- добавил страницы в папку pages и стили к каждой странице
- добавил вызовы каждого апи
- добавил экшены и редьюсеры для всех вызовов апи
- добавил modalswitch для двух режимов работы окна с ингредиентом - как модалка или как отдельная страница
- добавил роутинг
- добавил ограничение на авторизированных и не авторизированных пользователей на страницы
- добавил регистрацию, логин, сброс и восстановление пароля, обновление данных юзера

andryxxa93 commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -74,29 +77,41 @@ function BurgerConstructor() {
+                setModal(true);
+            }
+        }
+        // if ((firstIngredient.length>0)||(components.length>0)){
```
**Можно улучшить**
Ненужный код лучше удалять

andryxxa93 reviewed (changes requested):
Отличная работа! Всё работает как следует. Вы хорошо разобрались с роутами и авторизацией. 

Но есть пару замечаний, которые необходимо исправить.

**Необходимо исправить**
1. Необходимо, чтобы при прямом переходе на маршрут с id ингредиента открывалась страница с информацией об ингредиенте, а не модальное окно с ним.
2. При изменении данных профиля происходит ошибка. Данных в запросе не хватает, хотя они есть в инпутах - 
<img width="1228" alt="Снимок экрана 2022-11-01 в 22 00 15" src="https://user-images.githubusercontent.com/60173362/199316374-e701fa43-86b1-4885-bf55-c9cdcc4d153d.png">
3. Кнопки "Сохранить" и "Отмена" должны становиться активными только после того пользователь внес какие-то изменения в форму данных профиля.

4. При запросах использующих accessToken должен проверятся ответ сервера. Если время жизни токена истекло - ошибка 403 (jwt expired), то выполняется запрос обновления токена и повторяется запрос, который не был выполнен успешно.

Удачных правок!

EldenLordCommander commented:
@andryxxa93 
Добрый вечер! Понял, спасибо, но я немного не понимаю об описанных проблемах. Не знаю, можно ли тут писать вопросы к вам, ревьюеру, но я бы хотел уточнить пару моментов, потому что у меня описанные замечания работают корректно. Или я не понимаю о чем речь:

1) Вот два режима отображения ингридиента: в модальном окне на странице приложения, либо если просто открыть ссылку в url - отдельная страничка с id конкретного ингридиента. Как я понял из листинга Влада и из задания, такая логика и отображение должно быть
![изображение](https://user-images.githubusercontent.com/12035340/199323992-0f89e746-37d6-4c7b-bc35-1ec75b816038.png)
![изображение](https://user-images.githubusercontent.com/12035340/199324067-3217ae24-e59c-4efc-8887-4c1a72e15e98.png)

3) И это у меня работает. Вот архив с записью работы приложения, я добавлял в код логику, чтобы кнопки изначально были скрыты, и появлялись только если введенные данные в инпуте отличались от сохраненных в сторе.
[React App — Mozilla Firefox 2022-11-01 22-47-19.zip](https://github.com/EldenLordCommander/react-burger/files/9913558/React.App.Mozilla.Firefox.2022-11-01.22-47-19.zip)

4) в burger-api у меня есть метод fetchWithRefresh, который как раз выполняет данную задачу. Или речь о том, что у меня не везде где используется accessToken используется нужная логика? Не могли бы уточнить, пожалуйста?

andryxxa93 commented:
> @andryxxa93 Добрый вечер! Понял, спасибо, но я немного не понимаю об описанных проблемах. Не знаю, можно ли тут писать вопросы к вам, ревьюеру, но я бы хотел уточнить пару моментов, потому что у меня описанные замечания работают корректно. Или я не понимаю о чем речь:
> 
> 1. Вот два режима отображения ингридиента: в модальном окне на странице приложения, либо если просто открыть ссылку в url - отдельная страничка с id конкретного ингридиента. Как я понял из листинга Влада и из задания, такая логика и отображение должно быть
>    ![изображение](https://user-images.githubusercontent.com/12035340/199323992-0f89e746-37d6-4c7b-bc35-1ec75b816038.png)
>    ![изображение](https://user-images.githubusercontent.com/12035340/199324067-3217ae24-e59c-4efc-8887-4c1a72e15e98.png)
> 2. И это у меня работает. Вот архив с записью работы приложения, я добавлял в код логику, чтобы кнопки изначально были скрыты, и появлялись только если введенные данные в инпуте отличались от сохраненных в сторе.
>    [React App — Mozilla Firefox 2022-11-01 22-47-19.zip](https://github.com/EldenLordCommander/react-burger/files/9913558/React.App.Mozilla.Firefox.2022-11-01.22-47-19.zip)
> 3. в burger-api у меня есть метод fetchWithRefresh, который как раз выполняет данную задачу. Или речь о том, что у меня не везде где используется accessToken используется нужная логика? Не могли бы уточнить, пожалуйста?

1. По этому вопроса да, действительно всё верно реализовано.
2. Вот видео где вы можете увидеть, что кнопки становятся активными сразу же как только пользователь входит на страницу профиля и ещё ничего не успел изменить.  

https://user-images.githubusercontent.com/60173362/199435825-208e5fd4-fda9-4265-a225-6caae3e88139.mov

3. Вот сейчас если подождать 20 минут и попробовать обновить данные пользователя, то получаем ошибку. Она означает, что токен имеет неверное значение. 

https://user-images.githubusercontent.com/60173362/199435409-dd006274-317e-4059-a96e-1d20caa3fa51.mov

![Снимок экрана 2022-11-02 в 11 15 34](https://user-images.githubusercontent.com/60173362/199435682-d2021086-83ef-4f17-ae0c-0099f142d094.png)

EldenLordCommander commented:
> > @andryxxa93 Добрый вечер! Понял, спасибо, но я немного не понимаю об описанных проблемах. Не знаю, можно ли тут писать вопросы к вам, ревьюеру, но я бы хотел уточнить пару моментов, потому что у меня описанные замечания работают корректно. Или я не понимаю о чем речь:
> > 
> > 1. Вот два режима отображения ингридиента: в модальном окне на странице приложения, либо если просто открыть ссылку в url - отдельная страничка с id конкретного ингридиента. Как я понял из листинга Влада и из задания, такая логика и отображение должно быть
> >    ![изображение](https://user-images.githubusercontent.com/12035340/199323992-0f89e746-37d6-4c7b-bc35-1ec75b816038.png)
> >    ![изображение](https://user-images.githubusercontent.com/12035340/199324067-3217ae24-e59c-4efc-8887-4c1a72e15e98.png)
> > 2. И это у меня работает. Вот архив с записью работы приложения, я добавлял в код логику, чтобы кнопки изначально были скрыты, и появлялись только если введенные данные в инпуте отличались от сохраненных в сторе.
> >    [React App — Mozilla Firefox 2022-11-01 22-47-19.zip](https://github.com/EldenLordCommander/react-burger/files/9913558/React.App.Mozilla.Firefox.2022-11-01.22-47-19.zip)
> > 3. в burger-api у меня есть метод fetchWithRefresh, который как раз выполняет данную задачу. Или речь о том, что у меня не везде где используется accessToken используется нужная логика? Не могли бы уточнить, пожалуйста?
> 
>     1. По этому вопроса да, действительно всё верно реализовано.
> 
>     2. Вот видео где вы можете увидеть, что кнопки становятся активными сразу же как только пользователь входит на страницу профиля и ещё ничего не успел изменить.
> 
> 
> 2022-11-02.10.41.22.mov
> 
>     3. Вот сейчас если подождать 20 минут и попробовать обновить данные пользователя, то получаем ошибку. Она означает, что токен имеет неверное значение.
> 
> 
> 2022-11-02.11.13.27.mov
> 
> ![Снимок экрана 2022-11-02 в 11 15 34](https://user-images.githubusercontent.com/60173362/199435682-d2021086-83ef-4f17-ae0c-0099f142d094.png)

@andryxxa93 
2) У вас средствами браузера автоматически подставляется сохраненный браузером пароль в поле пароля. В моей реализации оно пустое и любое значение в этом поле считается изменением. Как видите, при открытии формы, имя и логин заполнены сразу - я их получаю и храню в сторе Редакса. А пароль подтягивается сам - это делает браузер, и после этого кнопки Отмена/Сохранить появляются (изначально они скрыты). 

Это сделано из-за того, что пароль я нигде не храню, поэтому пользователь может ввести любое значение в поле и отправить на апдейт. А может не слать, тогда пароль не изменится. 

Мне показалось такое поведение логичным, я встречал его на различных сайтах. Если оно не подходит, то я буду хранить пароль в сторе и поправлю, но мне кажется это не корректно. 

По остальным пунктам понял, доработаю, спасибо за ответ!

EldenLordCommander committed:
fix ошибок по ревью

- исправил проверку и обновление истекшего токена при обновлении пользователя
- поправил проблему, при которой на апдейт юзера не заполнялись поля для запроса
- исправил отображение кнопок Отмена/Сохранить

EldenLordCommander commented:
Исправил ошибки по ревью:
- исправил проверку и обновление истекшего токена при обновлении пользователя. У меня её не было, действительно. Проверил обновление через 20 минут - токен протух, выдал ошибку, запросил новый токен и обновил пользователя
- поправил проблему, при которой на апдейт юзера не заполнялись поля для запроса
- исправил отображение кнопок Отмена/Сохранить - проблема действительно была связана с функцией автозаполнения поля пароль у браузера. Добавил дополнительный аттрибут в поле пароль на форме пользователя, теперь в неё не должно подставлять сохраненное браузером значение, что вызывало реакцию кнопок "Отмена"/"Сохранить". Если что-то ввести в поле пароля - тогда триггер сработает. Значение пароля по умолчанию отправляется пустым на обновление, тем самым не обновляя его у пользователя

andryxxa93 reviewed (approved):
Супер, теперь всё работает как надо. Удачного продолжения обучения!
